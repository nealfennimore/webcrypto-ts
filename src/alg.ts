export namespace AES {
    export enum Mode {
        AES_CBC = "AES-CBC",
        AES_CTR = "AES-CTR",
        AES_GCM = "AES-GCM",
        AES_KW = "AES-KW",
    }

    export type Modes = `${Mode}`;
}

export namespace SHA {
    export enum Variant {
        SHA_1 = "SHA-1",
        SHA_256 = "SHA-256",
        SHA_384 = "SHA-384",
        SHA_512 = "SHA-512",
    }
    export type Variants = `${Variant}`;
    export type SecureVariants = `${Exclude<Variant, Variant.SHA_1>}`;
}

export namespace RSA {
    export enum Variant {
        RSA_OAEP = "RSA-OAEP",
        RSA_PSS = "RSA-PSS",
        RSASSA_PKCS1_v1_5 = "RSASSA-PKCS1-v1_5",
    }
    export type Variants = `${Variant}`;
}

export namespace EC {
    export enum Variant {
        ECDSA = "ECDSA",
        ECDH = "ECDH",
    }
    export type Variants = `${Variant}`;

    export enum Curve {
        P_256 = "P-256",
        P_384 = "P-384",
        P_521 = "P-521",
    }

    export type Curves = `${Curve}`;
}

export namespace Authentication {
    export enum Code {
        HMAC = "HMAC",
    }
    export type Codes = `${Code}`;
}
export namespace KDF {
    export enum Variant {
        PBKDF2 = "PBKDF2",
        HKDF = "HKDF",
    }
    export type Variants = `${Variant}`;
}
