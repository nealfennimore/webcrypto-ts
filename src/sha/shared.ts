export interface Sha1ArrayBuffer extends ArrayBuffer {
    _sha1ArrayBufferBrand: any;
}
export interface Sha256ArrayBuffer extends ArrayBuffer {
    _sha256ArrayBufferBrand: any;
}
export interface Sha384ArrayBuffer extends ArrayBuffer {
    _sha384ArrayBufferBrand: any;
}
export interface Sha512ArrayBuffer extends ArrayBuffer {
    _sha512ArrayBufferBrand: any;
}

export type ShaArrayBuffers =
    | Sha1ArrayBuffer
    | Sha256ArrayBuffer
    | Sha384ArrayBuffer
    | Sha512ArrayBuffer;

export namespace ShaShared {
    export const hexify = (digest: ShaArrayBuffers) =>
        Array.from(new Uint8Array(digest))
            .map((b) => b.toString(16).padStart(2, "0"))
            .join("");
}
