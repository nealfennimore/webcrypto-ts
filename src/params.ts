import * as alg from "./alg";

export type AesBlockSize = 128 | 192 | 256;

export interface EnforcedRsaHashedKeyGenParams extends RsaHashedKeyGenParams {
    name: alg.RSA.Variants;
    hash: alg.SHA.SecureVariants;
    modulusLength: 2048 | 4096;
}

export interface EnforcedEcKeyGenParams extends EcKeyGenParams {
    name: alg.EC.Variants;
    namedCurve: alg.EC.Curves;
}

export interface EnforcedAesKeyGenParams extends AesKeyGenParams {
    name: alg.AES.Modes;
    length: AesBlockSize;
}

export interface EnforcedHmacKeyGenParams extends HmacKeyGenParams {
    name: alg.Authentication.Code.HMAC;
    hash: alg.SHA.SecureVariants;
    length?: 512 | 1024;
}

export interface EnforcedRsaHashedImportParams extends RsaHashedImportParams {
    name: alg.RSA.Variants;
    hash: alg.SHA.SecureVariants;
}

export interface EnforcedEcKeyImportParams extends EcKeyImportParams {
    name: alg.EC.Variants;
    namedCurve: alg.EC.Curves;
}

export interface EnforcedHmacImportParams extends HmacImportParams {
    name: alg.Authentication.Code.HMAC;
    hash: alg.SHA.SecureVariants;
    length?: 512 | 1024;
}

export interface EnforcedRsaOaepParams extends RsaOaepParams {
    name: alg.RSA.Variant.RSA_OAEP;
}

export interface EnforcedRsaPssParams extends RsaPssParams {
    name: alg.RSA.Variant.RSA_PSS;
}
export interface EnforcedRsassaPkcs1v15Params extends Algorithm {
    name: alg.RSA.Variant.RSASSA_PKCS1_v1_5;
}

export interface EnforcedEcdsaParams extends EcdsaParams {
    name: alg.EC.Variant.ECDSA;
    hash: alg.SHA.SecureVariants;
}

export interface EnforcedEcdhKeyDeriveParams extends EcdhKeyDeriveParams {
    name: alg.EC.Variant.ECDH;
}

export interface AesGcmKeyAlgorithm extends AesKeyAlgorithm {
    name: alg.AES.Mode.AES_GCM;
    length: AesBlockSize;
}

export interface EnforcedAesGcmParams extends AesGcmParams {
    tagLength?: 32 | 64 | 96 | 104 | 112 | 120 | 128;
}

export interface AesCtrKeyAlgorithm extends AesKeyAlgorithm {
    name: alg.AES.Mode.AES_CTR;
    length: AesBlockSize;
}

export interface EnforcedAesCtrParams extends AesCtrParams {
    name: alg.AES.Mode.AES_CTR;
}

export interface AesCbcKeyAlgorithm extends AesKeyAlgorithm {
    name: alg.AES.Mode.AES_CBC;
    length: AesBlockSize;
}

export interface EnforcedAesCbcParams extends AesCbcParams {
    name: alg.AES.Mode.AES_CBC;
}

export interface AesKwKeyAlgorithm extends KeyAlgorithm {
    name: alg.AES.Mode.AES_KW;
}

export interface EnforcedAesKwParams {
    name: alg.AES.Mode.AES_KW;
}

export interface EnforcedHkdfParams extends HkdfParams {
    name: alg.KDF.Variant.HKDF;
    hash: alg.SHA.SecureVariants;
}
export interface EnforcedPbkdf2Params extends Pbkdf2Params {
    name: alg.KDF.Variant.PBKDF2;
    hash: alg.SHA.SecureVariants;
    iterations: 720_000 | 310_000 | 310_000 | 120_000;
}

export type EnforcedAesKeyAlgorithms =
    | AesGcmKeyAlgorithm
    | AesCtrKeyAlgorithm
    | AesCbcKeyAlgorithm
    | AesKwKeyAlgorithm;
export type EnforcedAesKeyAlgorithmNames = Pick<
    EnforcedAesKeyAlgorithms,
    "name"
>;
export type EnforcedAesParams =
    | EnforcedAesGcmParams
    | EnforcedAesCtrParams
    | EnforcedAesCbcParams
    | EnforcedAesKwParams;
export type EnforcedImportParams =
    | EnforcedRsaHashedImportParams
    | EnforcedEcKeyImportParams
    | EnforcedHmacImportParams
    | EnforcedAesKeyAlgorithmNames;
export type EnforcedKeyGenParams =
    | EnforcedRsaHashedKeyGenParams
    | EnforcedEcKeyGenParams
    | EnforcedHmacKeyGenParams
    | EnforcedAesKeyGenParams;
export type EnforcedKeyDeriveParams =
    | EnforcedEcdhKeyDeriveParams
    | EnforcedHkdfParams
    | EnforcedPbkdf2Params;
