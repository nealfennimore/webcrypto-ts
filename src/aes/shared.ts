/**
 * Shared code for AES
 * @module
 */

import { getKeyUsagePairsByAlg } from "../key_usages.js";
import * as params from "../params.js";
import * as WebCrypto from "../webcrypto.js";

export interface AesGcmCryptoKey extends CryptoKey {
    _aesGcmKeyBrand: any;
}

export interface AesKwCryptoKey extends CryptoKey {
    _aesKwKeyBrand: any;
}

export interface AesCtrCryptoKey extends CryptoKey {
    _aesCtrKeyBrand: any;
}

export interface AesCbcCryptoKey extends CryptoKey {
    _aesCbcKeyBrand: any;
}

export type AesCryptoKeys =
    | AesCbcCryptoKey
    | AesKwCryptoKey
    | AesGcmCryptoKey
    | AesCtrCryptoKey;

export namespace Alg {
    export enum Mode {
        AES_CBC = "AES-CBC",
        AES_CTR = "AES-CTR",
        AES_GCM = "AES-GCM",
        AES_KW = "AES-KW",
    }

    export type Modes = `${Mode}`;
}

export namespace AesShared {
    export async function generateKey<T extends CryptoKey>(
        algorithm: params.EnforcedAesKeyGenParams,
        extractable: boolean = true,
        keyUsages?: KeyUsage[]
    ): Promise<T> {
        return await WebCrypto.generateKey<T, params.EnforcedAesKeyGenParams>(
            algorithm,
            extractable,
            keyUsages ?? getKeyUsagePairsByAlg(algorithm.name)
        );
    }

    export async function importKey<T extends CryptoKey>(
        format: KeyFormat,
        key: BufferSource | JsonWebKey,
        algorithm: params.EnforcedAesKeyAlgorithms,
        extractable: boolean = true,
        keyUsages?: KeyUsage[]
    ): Promise<T> {
        return await WebCrypto.importKey<T, params.EnforcedAesKeyAlgorithms>(
            format as any,
            key as any,
            algorithm,
            extractable,
            keyUsages ?? getKeyUsagePairsByAlg(algorithm.name)
        );
    }

    export async function exportKey<T extends CryptoKey>(
        format: KeyFormat,
        key: T
    ): Promise<JsonWebKey | ArrayBuffer> {
        return await WebCrypto.exportKey(format as any, key);
    }

    export async function encrypt<T extends CryptoKey>(
        algorithm: Exclude<
            params.EnforcedAesParams,
            params.EnforcedAesKwParams
        >,
        key: T,
        data: BufferSource
    ): Promise<ArrayBuffer> {
        return await WebCrypto.encrypt(algorithm, key, data);
    }

    export async function decrypt<T extends CryptoKey>(
        algorithm: Exclude<
            params.EnforcedAesParams,
            params.EnforcedAesKwParams
        >,
        key: T,
        data: BufferSource
    ): Promise<ArrayBuffer> {
        return await WebCrypto.decrypt(algorithm, key, data);
    }

    export async function wrapKey<T extends CryptoKey>(
        format: KeyFormat,
        key: CryptoKey,
        wrappingkey: T,
        wrapAlgorithm: params.EnforcedAesParams
    ): Promise<ArrayBuffer> {
        return await WebCrypto.wrapKey(
            format as any,
            key,
            wrappingkey,
            wrapAlgorithm
        );
    }
    export async function unwrapKey<T extends CryptoKey>(
        format: KeyFormat,
        wrappedKey: BufferSource,
        wrappedKeyAlgorithm: params.EnforcedImportParams,
        unwrappingKey: T,
        unwrappingKeyAlgorithm: params.EnforcedAesParams,
        extractable: boolean = true,
        keyUsages?: KeyUsage[]
    ): Promise<CryptoKey> {
        return await WebCrypto.unwrapKey(
            format as any,
            wrappedKey,
            unwrappingKey,
            unwrappingKeyAlgorithm,
            wrappedKeyAlgorithm,
            extractable,
            keyUsages ?? getKeyUsagePairsByAlg(wrappedKeyAlgorithm.name)
        );
    }
}
