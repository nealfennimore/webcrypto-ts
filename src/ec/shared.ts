import { WebCrypto } from "../crypto";
import { getKeyUsagePairsByAlg } from "../keyUsages";
import * as params from "../params";

export interface EcKey extends CryptoKey {}
export interface EcKeyPair extends CryptoKeyPair {}

export namespace SharedEc {
    export async function generateKey(
        algorithm: params.EnforcedEcKeyGenParams,
        extractable: boolean = true,
        keyUsages?: KeyUsage[]
    ): Promise<EcKeyPair> {
        return await WebCrypto.generateKey<
            EcKeyPair,
            params.EnforcedEcKeyGenParams
        >(
            algorithm,
            extractable,
            keyUsages ?? getKeyUsagePairsByAlg(algorithm.name)
        );
    }

    export async function importKey(
        format: KeyFormat,
        keyData: BufferSource | JsonWebKey,
        algorithm: params.EnforcedEcKeyImportParams,
        extractable: boolean = true,
        keyUsages?: KeyUsage[]
    ): Promise<EcKey> {
        return await WebCrypto.importKey<
            EcKey,
            params.EnforcedEcKeyImportParams
        >(
            format as any,
            keyData as any,
            algorithm,
            extractable,
            keyUsages ?? getKeyUsagePairsByAlg(algorithm.name)
        );
    }

    export async function exportKey(
        format: KeyFormat,
        keyData: EcKey
    ): Promise<JsonWebKey | ArrayBuffer> {
        return await WebCrypto.exportKey(format as any, keyData);
    }
}
