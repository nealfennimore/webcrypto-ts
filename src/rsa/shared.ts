import { WebCrypto } from "../crypto";
import { getKeyUsagePairsByAlg } from "../keyUsages";
import * as params from "../params";

export interface RsaKey extends CryptoKey {}
export interface RsaKeyPair extends CryptoKeyPair {}

export namespace RsaShared {
    export async function generateKey(
        algorithm: params.EnforcedRsaHashedKeyGenParams,
        extractable: boolean = true,
        keyUsages?: KeyUsage[]
    ): Promise<RsaKey | RsaKeyPair> {
        return await WebCrypto.generateKey<
            RsaKey | RsaKeyPair,
            params.EnforcedRsaHashedKeyGenParams
        >(
            algorithm,
            extractable,
            keyUsages ?? getKeyUsagePairsByAlg(algorithm.name)
        );
    }

    export async function importKey(
        format: KeyFormat,
        algorithm: params.EnforcedRsaHashedImportParams,
        keyData: BufferSource | JsonWebKey,
        extractable: boolean = true,
        keyUsages?: KeyUsage[]
    ): Promise<RsaKey> {
        return await WebCrypto.importKey<
            RsaKey,
            params.EnforcedRsaHashedImportParams
        >(
            format as any,
            keyData as any,
            algorithm,
            extractable,
            keyUsages ?? getKeyUsagePairsByAlg(algorithm.name)
        );
    }

    export async function exportKey(
        format: KeyFormat,
        keyData: RsaKey
    ): Promise<JsonWebKey | ArrayBuffer> {
        return await WebCrypto.exportKey(format as any, keyData);
    }

    export async function sign(
        algorithm:
            | params.EnforcedRsaPssParams
            | params.EnforcedRsassaPkcs1v15Params,
        keyData: RsaKey,
        data: BufferSource
    ): Promise<ArrayBuffer> {
        return await WebCrypto.sign(algorithm, keyData, data);
    }

    export async function verify(
        algorithm:
            | params.EnforcedRsaPssParams
            | params.EnforcedRsassaPkcs1v15Params,
        keyData: RsaKey,
        signature: BufferSource,
        data: BufferSource
    ): Promise<boolean> {
        return await WebCrypto.verify(algorithm, keyData, signature, data);
    }
}
