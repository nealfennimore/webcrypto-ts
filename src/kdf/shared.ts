/**
 * Shared code for KDF
 * @module
 */
import type { AesCryptoKeys } from "../aes/index.js";
import { AesProxiedCryptoKeys } from "../aes/shared.js";
import { HmacCryptoKey, HmacProxiedCryptoKey } from "../hmac/index.js";
import { DeriveKeyUsagePair, getKeyUsagePairsByAlg } from "../key_usages.js";
import * as params from "../params.js";
import * as proxy from "../proxy.js";
import * as WebCrypto from "../webcrypto.js";

export interface Pbkdf2KeyMaterial extends CryptoKey {
    _pbkdf2KeyMaterialBrand: any;
}
export interface HkdfKeyMaterial extends CryptoKey {
    _kkdfKeyMaterialBrand: any;
}

export namespace Alg {
    export enum Variant {
        PBKDF2 = "PBKDF2",
        HKDF = "HKDF",
    }
    export type Variants = `${Variant}`;
}

export namespace KdfShared {
    export async function generateKeyMaterial<K extends CryptoKey>(
        format: KeyFormat,
        key: BufferSource,
        algorithm: Alg.Variants,
        extractable: boolean = false
    ): Promise<K> {
        return await WebCrypto.importKey(
            format as any,
            key,
            algorithm,
            extractable,
            DeriveKeyUsagePair
        );
    }

    export async function deriveKey(
        algorithm: params.EnforcedPbkdf2Params | params.EnforcedHkdfParams,
        baseKey: Pbkdf2KeyMaterial | HkdfKeyMaterial,
        derivedKeyType:
            | params.EnforcedAesKeyGenParams
            | params.EnforcedHmacKeyGenParams,
        extractable: boolean = true,
        keyUsages?: KeyUsage[]
    ): Promise<AesCryptoKeys | HmacCryptoKey> {
        return await WebCrypto.deriveKey<
            AesCryptoKeys | HmacCryptoKey,
            params.EnforcedAesKeyGenParams | params.EnforcedHmacKeyGenParams
        >(
            algorithm,
            baseKey,
            derivedKeyType,
            extractable,
            keyUsages ?? getKeyUsagePairsByAlg(derivedKeyType.name)
        );
    }

    export async function deriveBits(
        algorithm: params.EnforcedPbkdf2Params | params.EnforcedHkdfParams,
        baseKey: Pbkdf2KeyMaterial | HkdfKeyMaterial,
        length: number
    ): Promise<ArrayBuffer> {
        return await WebCrypto.deriveBits(algorithm, baseKey, length);
    }
}

export interface HkdfProxiedKeyMaterial
    extends proxy.ProxiedCryptoKey<HkdfKeyMaterial> {
    deriveKey(
        algorithm: Omit<params.EnforcedHkdfParams, "name">,
        derivedKeyType:
            | params.EnforcedAesKeyGenParams
            | params.EnforcedHmacKeyGenParams,
        extractable?: boolean,
        keyUsages?: KeyUsage[]
    ): Promise<AesProxiedCryptoKeys | HmacProxiedCryptoKey>;

    deriveBits(
        algorithm: Omit<params.EnforcedHkdfParams, "name">,
        length: number
    ): Promise<ArrayBuffer>;
    exportKey: (format: KeyFormat) => Promise<JsonWebKey | ArrayBuffer>;
}

export interface Pbkdf2ProxiedKeyMaterial
    extends proxy.ProxiedCryptoKey<Pbkdf2KeyMaterial> {
    deriveKey(
        algorithm: Omit<params.EnforcedPbkdf2Params, "name" | "iterations">,
        derivedKeyType:
            | params.EnforcedAesKeyGenParams
            | params.EnforcedHmacKeyGenParams,
        extractable?: boolean,
        keyUsages?: KeyUsage[]
    ): Promise<AesProxiedCryptoKeys | HmacProxiedCryptoKey>;
    deriveBits(
        algorithm: Omit<params.EnforcedPbkdf2Params, "name" | "iterations">,
        length: number
    ): Promise<ArrayBuffer>;
    exportKey: (format: KeyFormat) => Promise<JsonWebKey | ArrayBuffer>;
}
