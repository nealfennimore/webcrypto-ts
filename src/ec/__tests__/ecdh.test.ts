import * as alg from "../../alg.js";
import * as params from "../../params.js";
import { EC } from "../index.js";
import { EcdhCryptoKeyPair } from "../shared.js";

const { ECDH } = EC;

describe("ECDH", () => {
    let keyPair: EcdhCryptoKeyPair;
    beforeEach(async () => {
        keyPair = await ECDH.generateKey();
    });
    it("should generate key", async () => {
        expect(keyPair).toMatchSnapshot();
    });
    it("should import and export key", async () => {
        let jwk = await ECDH.exportKey("jwk", keyPair.publicKey);
        const importedPubKey = await ECDH.importKey(
            "jwk",
            jwk,
            { namedCurve: "P-521" },
            true,
            []
        );

        expect(await ECDH.exportKey("jwk", importedPubKey)).toEqual(jwk);

        jwk = await ECDH.exportKey("jwk", keyPair.privateKey);
        const importedPrivKey = await ECDH.importKey("jwk", jwk, {
            namedCurve: "P-521",
        });

        expect(await ECDH.exportKey("jwk", importedPrivKey)).toEqual(jwk);
    });
    it("should derive bits", async () => {
        const otherKeyPair = await ECDH.generateKey();

        const bits = await ECDH.deriveBits(
            { public: otherKeyPair.publicKey },
            keyPair.privateKey,
            128
        );
        expect(bits.byteLength).toEqual(16);

        await expect(
            ECDH.deriveBits(
                { public: otherKeyPair.publicKey },
                keyPair.privateKey,
                127
            )
        ).rejects.toThrowError(RangeError);
    });
    it("should derive keys", async () => {
        const otherKeyPair = await ECDH.generateKey();
        const hmacParams: params.EnforcedHmacKeyGenParams = {
            name: alg.Authentication.Code.HMAC,
            hash: alg.SHA.Variant.SHA_512,
            length: 512,
        };
        let key = await ECDH.deriveKey(
            { public: otherKeyPair.publicKey },
            keyPair.privateKey,
            hmacParams
        );
        expect(key).toMatchSnapshot();
    });
});
