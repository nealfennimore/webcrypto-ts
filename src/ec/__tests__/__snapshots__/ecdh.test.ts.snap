// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ECDH should derive keys 1`] = `
CryptoKey {
  Symbol(kKeyObject): SecretKeyObject {
    Symbol(kKeyType): "secret",
  },
  Symbol(kAlgorithm): Object {
    "hash": Object {
      "hash": undefined,
      "name": "SHA-512",
    },
    "length": 512,
    "name": "HMAC",
  },
  Symbol(kExtractable): true,
  Symbol(kKeyUsages): Array [
    "sign",
    "verify",
  ],
}
`;

exports[`ECDH should generate key 1`] = `
Object {
  "privateKey": CryptoKey {
    Symbol(kKeyObject): PrivateKeyObject {
      Symbol(kKeyType): "private",
    },
    Symbol(kAlgorithm): Object {
      "name": "ECDH",
      "namedCurve": "P-521",
    },
    Symbol(kExtractable): true,
    Symbol(kKeyUsages): Array [
      "deriveKey",
      "deriveBits",
    ],
  },
  "publicKey": CryptoKey {
    Symbol(kKeyObject): PublicKeyObject {
      Symbol(kKeyType): "public",
    },
    Symbol(kAlgorithm): Object {
      "name": "ECDH",
      "namedCurve": "P-521",
    },
    Symbol(kExtractable): true,
    Symbol(kKeyUsages): Array [],
  },
}
`;
